### DISCUSSION TITLE
set_position 持仓不同步

### DISCUSSION THREAD

**Discussion by quant1box** [wondertrader/wondertrader#59] [Q&A] [answered]
版本——0.9.2

问题描述：做了一个和数据库商品持仓列表同步的程序，策略中使用交易指令为set_position,但是并不能起到持仓同步的。检查原因是目标持仓从数据库获取是正确的，但是获取当前持仓的数据是不对的，不是正真账户的当前持仓。这个当前持仓是怎么维护的？是本地csv或者json维护的，还是直接从接口取得的？在本地那么可以自己手工去修改以达到目标持仓的效果？

**Comment by wondertrader** [comment]
请贴一下你的代码

**Reply by quant1box** [reply]
1. ctp配置文件
![image](https://user-images.githubusercontent.com/19401416/162666944-10f1a23c-3868-46a7-a0c3-a04ad83e348b.png)

2 执行器配置文件
![image](https://user-images.githubusercontent.com/19401416/162667032-6914c046-ea97-49c9-8271-5f8daed46e52.png)

3 策略代码仓位执行模块
![image](https://user-images.githubusercontent.com/19401416/162667122-21eb724d-1dc0-47ec-91cf-7bd94bdc741a.png)

4 我本来应该截图保存在本地的策略持仓和通道账户实际持仓做对比的，但是我上午手工把多余的持仓平掉了。我简单描述一下具体情况：
day1交易日实际持仓：{"a":1,"b":-1,"c":3} 
day2交易日目标持仓：{”b“:-1,"c":3,"d":-3}
但是在day2执行持仓同步程序以后账户的实际持仓为：{”a“:1,"b":-1,"c":3，”d“:-3}, 本地持仓文件和目标持仓文件一致，但是和通道 账户实际持仓不一样

**Reply by wondertrader** [reply]
你贴的这段逻辑是我提供的，而且是我用在实盘的，没有问题的
我建议你打印出来对一下：
1、数据库读到的目标仓位
2、合并之前的本地持仓
3、合并之后的最终持仓

**Comment by wondertrader** [comment, marked as answer]
WonderTrader的执行器有一个管理机制：**不在目标仓位中的合约不管理**
按照你问题中的例子：
day1交易日目标持仓：{"a":1,"b":-1,"c":3}
day2交易日目标持仓：{”b“:-1,"c":3,"d":-3}
最终处理过以后，如下：
day2交易日最终目标持仓：{"a":0,”b“:-1,"c":3,"d":-3}
执行器按照这个目标仓位进行执行
而同步之前交易通道中的持仓如下：
{"a":1,”b“:-1,"c":3,"d":-3,"e":-2}
执行器就会进行以下操作：
a：平多1手
b：保持不变
c：保持不变
d：开空3手
e：忽略（注意和保持不变的区别）
以上的逻辑概括起来就是：**如果交易通道中有不在管理中的仓位，执行器不会进行任何调整**
